libelement/src/log_errors.hpp

```diff
@@ -196,27 +168,349 @@ namespace element
+};
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+    element::log_error_message_code::parse_identifier_failed,
+    element_result::ELEMENT_ERROR_INVALID_IDENTIFIER,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "invalid identifier '{}'",
+    std::string
+);
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+   element::log_error_message_code::parse_identifier_reserved,
+    element_result::ELEMENT_ERROR_RESERVED_IDENTIFIER,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "identifier '{}' is reserved",
+    std::string
+);
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+   element::log_error_message_code::parse_typename_not_identifier,
+    element_result::ELEMENT_ERROR_INVALID_TYPENAME,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "found '{}' when a type was expected."
+    "\nnote: a type must be an identifier or a chain of them, such as 'MyNamespace.MyStruct'.",
+    std::string
+);
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+   element::log_error_message_code::parse_port_failed,
+    element_result::ELEMENT_ERROR_INVALID_PORT,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "found '{}' when a port was expected."
+    "\nnote: a port must be either an identifier or an underscore.",
+    std::string
+);
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+   element::log_error_message_code::parse_exprlist_empty,
+    element_result::ELEMENT_ERROR_MISSING_CONTENTS_FOR_CALL,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "expected to find something in the contents of the call to '{}', but nothing is contained within the parenthesis."
+    "\nnote: perhaps you meant to call a function with no arguments, in which case you must omit the parenthesis.",
+    std::string
+);
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+   element::log_error_message_code::parse_exprlist_missing_closing_parenthesis,
+    element_result::ELEMENT_ERROR_MISSING_PARENTHESIS_FOR_CALL,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "expected to find a ')' at the end of the call to '{}', but found '{}' instead.",
+    std::string, std::string
+);
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+   element::log_error_message_code::parse_call_invalid_expression,
+    element_result::ELEMENT_ERROR_INVALID_CONTENTS_FOR_CALL,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "expected to find an identifier or number in the contents of the call to '{}', but found '{}' instead.",
+    std::string, std::string
+);
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+   element::log_error_message_code::parse_expression_failed,
+    element_result::ELEMENT_ERROR_INVALID_EXPRESSION,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "failed to parse the expression '{}'."
+    "\nnote: it must start with an identifier, an underscore, or a number.",
+    std::string
+);
+
+MAKE_LOG_ERROR_MESSAGE_INFO(
+   element::log_error_message_code::parse_declaration_invalid_identifier,
+    element_result::ELEMENT_ERROR_INVALID_IDENTIFIER,
+    element_stage::ELEMENT_STAGE_PARSER,
+    "failed to parse the expression '{}'."
+    "found '{}' when parsing a declaration, expected a valid identifier.",
+    std::string
```

Comment:
Should this be two `std::string`s per the message above? (and should the message above have a `\n` between the lines?)

Reply:
it definitely should 😁 weird that this wasn't picked up at compile-time, will need to see why and fix that flaw

Reply:
the reason it wasn't picked up is because the codebase isn't ever trying to create that error message, so none of the compile-time checks are made.

I've fixed that issue, and two others where we had `'{'` which fmt requires we delimit as `{{` (since it looks for `{}`). I've added `format_check` to the struct and a comment explaining it, not perfect, but it should handle most things correctly and we can modify it to ignore any problem cases that come up in the future.